### DISCUSSION TITLE
3-site non-Hamiltionian TEBD

### DISCUSSION THREAD

**Discussion by matijacell** [jcmgray/quimb#166] [Q&A]
Hi everyone,

I am working on a system with a non-Hamiltionian 3-site update rule (an 8x8 real matrix) in the bulk of the system and a 2 site update rule on the boundary (a 4x4 real matrix). Neighbouring sites commute with one another, so no Trotterization is required. I wrote my own TEBD algorithm which is painfully slow, and I wanted to use your faster framework, but I am unsure of how to implement a non-Hamiltionian model within QUIMB.

Should I simply create MPO representations of my local 3 site operators (and 2 site on the bulk) and I can then propagate my system in time using that? It feels like a question asked recently should be relevant (https://github.com/jcmgray/quimb/discussions/161).

If anyone has any suggestions, thank you very much in advance!

**Comment by jcmgray** [comment]
Hi @matijacell, I guess by update rule you mean you have essentially gates such that?
$$\prod_i G_i|\psi(t)\rangle = |\psi(t + \delta t)\rangle$$

There are a few ways to do it maybe:
1. The `MatrixProductState.gate` function, with `contract='swap+split'`. The problem here is that this would need some basic tweaks to support 3 body gates
2. Build the MPO, and simply apply and compress - this might be easiest way to go, but possibly slow if the terms are time-dependent or something.
3. 'Fit' the gate at each stage, which would involve something like:

    1. canonicalize around the sites
    2. construct the local TN of sites and gate $G$ (uncontracted in)
    3. view the sites without G, and fit to above object

   The problem here is you won't directly have adaptive bond dimensions
